File: 2025-04-April-GFG-POTD/April 26 - Is Binary Tree Heap/Solution.cpp
#include <iostream>
#include <sstream>
#include <queue>
#include <string>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

Node* buildTree(string str) {
    if (str.length() == 0 || str[0] == 'N') return nullptr;

    stringstream ss(str);
    string word;
    vector<string> ip;

    while (ss >> word) ip.push_back(word);

    Node* root = new Node(stoi(ip[0]));
    queue<Node*> q;
    q.push(root);

    int i = 1;
    while (!q.empty() && i < ip.size()) {
        Node* curr = q.front();
        q.pop();

        if (ip[i] != "N") {
            curr->left = new Node(stoi(ip[i]));
            q.push(curr->left);
        }
        i++;
        if (i >= ip.size()) break;

        if (ip[i] != "N") {
            curr->right = new Node(stoi(ip[i]));
            q.push(curr->right);
        }
        i++;
    }

    return root;
}

bool isHeap(Node* root) {
    if (!root) return true;

    queue<Node*> q;
    q.push(root);
    bool nullChildSeen = false;

    while (!q.empty()) {
        Node* curr = q.front();
        q.pop();

        if (curr->left) {
            if (nullChildSeen || curr->left->data > curr->data)
                return false;
            q.push(curr->left);
        } else {
            nullChildSeen = true;
        }

        if (curr->right) {
            if (nullChildSeen || curr->right->data > curr->data)
                return false;
            q.push(curr->right);
        } else {
            nullChildSeen = true;
        }
    }

    return true;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        string s;
        getline(cin, s);
        Node* root = buildTree(s);
        cout << (isHeap(root) ? "true" : "false") << endl;
        cout << "~" << endl;
    }

    return 0;
}

File: 2025-04-April-GFG-POTD/April 26 - Is Binary Tree Heap/Solution.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def buildTree(s):
    if not s or s[0] == "N":
        return None

    ip = s.strip().split()
    root = Node(int(ip[0]))
    q = deque([root])

    i = 1
    while q and i < len(ip):
        curr = q.popleft()

        if ip[i] != "N":
            curr.left = Node(int(ip[i]))
            q.append(curr.left)
        i += 1

        if i >= len(ip): break

        if ip[i] != "N":
            curr.right = Node(int(ip[i]))
            q.append(curr.right)
        i += 1

    return root

def isHeap(root):
    if not root:
        return True

    q = deque([root])
    nullChildSeen = False

    while q:
        curr = q.popleft()

        if curr.left:
            if nullChildSeen or curr.left.data > curr.data:
                return False
            q.append(curr.left)
        else:
            nullChildSeen = True

        if curr.right:
            if nullChildSeen or curr.right.data > curr.data:
                return False
            q.append(curr.right)
        else:
            nullChildSeen = True

    return True

# Driver code
t = int(input())
for _ in range(t):
    s = input()
    root = buildTree(s)
    print("true" if isHeap(root) else "false")
    print("~")
